### PR TITLE
feat(ticket#206): Edit User Details and Create Server Action for it 

### PR DESCRIPTION
**Pull request by Mubashar1009**
## Change Request

- Incharge or shift-incharge can update members details and after updating member details , they can  move to the member lists screen so that changes can be seen 
- Create api for updating member details 
- Create Server Action for updating member details

### Implementation

- Created edit route for updating specific member details 
- Created API for updating member details
- Created server action for getting member details 
- Use `MemberFormBuilder.tsx` for updating member details
- Created `updateUser` function for updating it 

### Resources
image
image

### Test

image

### CODE REVIEW

**Review by ghulamhafeez — APPROVED** [review]

File: src/app/members/actions/update-user.test.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      phoneNumber: "923030000726",
+      shift: Shift.ShiftB,
+      address: "Multan\nmultan",
+    });
+  });
+  it("Check Update user implementation ", async () => {
+    const mock = vi.fn().mockImplementation(updateUser);
+
+    await waitFor(() => {
+      mock(mockValues, {
```
Comment (Monfernape) [inline src/app/members/actions/update-user.test.tsx:56]: What's happening here?

File: src/app/members/actions/update-user.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+"use server";
+import { getSupabaseClient } from "@/utils/supabase/supabaseClient";
+import { UserBuilder } from "../components/MemberFormBuilder";
+import { Tables } from "@/lib/db";
+import { redirect } from "next/navigation";
+import { Routes } from "@/lib/routes";
+
+export const updateUser = async (values: UserBuilder, id: number) => {
+  const supabase = await getSupabaseClient();
+  const { error } = await supabase
```
Comment (Monfernape) [inline src/app/members/actions/update-user.ts:10]: We need to check the user role here to make sure not that only shift incharge and incharge can trigger this endpoint.

File: src/app/members/components/MemberActionRender.tsx [hunk 3]
```diff
@@ -15,13 +17,14 @@ type Props = {
...
-  const { role, status } = memberInfo;
+  const router = useRouter();
+  const { role, status, id } = memberInfo;
   const handleViewDetails = () => {
     return;
   };
 
   const handleEditInfo = () => {
-    return;
+    router.push(Routes.EditMember.replace("id", String(id)));
```
[the remark this answers is not in the document]
Comment (Monfernape) [reply src/app/members/components/MemberActionRender.tsx:27]: More better would be to not add `/` in the enum and add here. It's more readable

File: src/app/members/components/MemberFormBuilder.tsx [hunk 4]
```diff
@@ -114,14 +121,14 @@ export const MemberFormBuilder = () => {
...
     defaultValues: {
-      name: "",
-      phoneNumber: "",
-      date_of_birth: undefined,
-      cnic: "",
-      shift: Shift.ShiftA,
-      address: "",
-      role: UserRole.Member,
-      ehad_duration: new Date(),
+      name: member?.name ?? "",
```
Comment (Monfernape) [inline src/app/members/components/MemberFormBuilder.tsx:124]: Nice

File: src/app/members/components/MemberFormBuilder.tsx [hunk 5]
```diff
@@ -132,14 +139,16 @@ export const MemberFormBuilder = () => {
...
 
       if (!result) {
-        form.reset();
         toast({
-          title: "User created successfully",
-          description: "You will receive message shortly",
+          title: member?.id
+            ? "User updated successfully"
+            : "User created successfully",
+          description: member?.id ? "" : "You will receive message shortly",
```
Comment (Monfernape) [inline src/app/members/components/MemberFormBuilder.tsx:151]: Description doesn't make sense. Leave it empty if not needed.

File: src/app/members/edit/[id]/page.tsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import { Tables } from "@/lib/db";
+
+type Parameters = {
+  params: {
+    id: string;
+  };
+};
+const page = async ({ params }: Parameters) => {
+  const { id } = params;
+  const supabase = await getSupabaseClient();
```
[the remark this answers is not in the document]
Comment (Monfernape) [reply src/app/members/edit/[id]/page.tsx:13]: Let's create an action.

File: src/app/members/edit/[id]/page.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import { MemberFormBuilder } from "../../components/MemberFormBuilder";
+import { getSupabaseClient } from "@/utils/supabase/supabaseClient";
+import { Tables } from "@/lib/db";
+
+type Parameters = {
+  params: {
+    id: string;
+  };
+};
+const page = async ({ params }: Parameters) => {
```
Comment (Monfernape) [inline src/app/members/edit/[id]/page.tsx:11]: Pascal case please.

**Review by Monfernape — CHANGES_REQUESTED** [review]
Seems like it's missing check for who can update the user.

File: src/app/members/components/MemberFormBuilder.tsx (hunk 5 above)
Comment (Mubashar1009) [reply src/app/members/components/MemberFormBuilder.tsx]: fixed

File: src/app/members/edit/[id]/page.tsx (hunk 7 above)
Comment (Mubashar1009) [reply src/app/members/edit/[id]/page.tsx]: fixed

**Review by Monfernape — APPROVED** [review]

**Pull request merged**